permissioned-token/program/src/lib.rs

```diff
@@ -0,0 +1,221 @@
+solana_program::declare_id!("PTxTEZXSadZ39at9G3hdXyYkKfyohTG3gCfNuSVnq4K");
+
+use borsh::BorshDeserialize;
+use solana_program::{
+    account_info::AccountInfo, entrypoint::ProgramResult, msg, program::invoke,
+    program_error::ProgramError, program_pack::Pack, pubkey::Pubkey, rent::Rent,
+    system_instruction, sysvar::Sysvar,
+};
+use spl_associated_token_account::instruction::create_associated_token_account;
+
+#[track_caller]
+#[inline(always)]
+pub fn assert_with_msg(v: bool, err: impl Into<ProgramError>, msg: &str) -> ProgramResult {
+    if v {
+        Ok(())
+    } else {
+        let caller = std::panic::Location::caller();
+        msg!("{}. \n{}", msg, caller);
+        Err(err.into())
+    }
+}
+
+pub mod accounts;
+pub mod instruction;
+pub mod token;
+use accounts::{Close, InitializeAccount, InitializeMint, MintOrBurn, Transfer};
+use instruction::PermissionedTokenInstruction;
+use token::{burn, close, freeze, initialize_mint, mint_to_signed, thaw, transfer};
+
+#[cfg(not(feature = "no-entrypoint"))]
+solana_program::entrypoint!(process_instruction);
+
+#[inline]
+fn get_freeze_authority_seeds_checked(
+    upstream_authority: &Pubkey,
+    expected_key: &Pubkey,
+) -> Result<Vec<Vec<u8>>, ProgramError> {
+    let (freeze_key, seeds) = get_freeze_authority(upstream_authority);
+    assert_with_msg(
+        expected_key == &freeze_key,
+        ProgramError::InvalidInstructionData,
+        "Invalid freeze authority",
+    )?;
+    Ok(seeds)
+}
+
+#[inline]
+fn get_freeze_authority(upstream_authority: &Pubkey) -> (Pubkey, Vec<Vec<u8>>) {
+    let mut seeds = vec![upstream_authority.as_ref().to_vec()];
```

Comment:
nit: it'd be better to separate the domain for the freeze authority by adding some additional seed like `b"freeze-authority"`

Reply:
You can still get into weird situations that go against the spirit of the two separate PDAs.  For example, if you generate a keypair, create a mint on it, and also use that keypair as the upstream authority, then your derived mint authority and freeze authorities will be the same address.

This isn't exploitable in this case, but it's still a source of confusion.  If you want to keep things simple, I'd recommend just having one PDA that mints and freezes.

Reply:
I'll make the change the unify the PDAs